itests/gas_estimation_test.go

```diff
@@ -52,3 +60,110 @@ func TestEstimateGasNoFunds(t *testing.T) {
 	require.NoError(t, err)
 	require.NotZero(t, limit)
 }
+
+// Make sure that we correctly calculate the inclusion cost. Especially, make sure the FVM and Lotus
+// agree and that:
+//  1. The FVM will never charge _less_ than the inclusion cost.
+//  2. The FVM will never fine a storage provider for including a message that costs exactly the
+//     inclusion cost.
+func TestEstimateInclusion(t *testing.T) {
+	ctx := context.Background()
+
+	kit.QuietMiningLogs()
+
+	// We need this to be "correct" in this test so that lotus can get the correct gas value
+	// (which, unfortunately, looks at the height and not the current network version).
+	oldPrices := vm.Prices
+	vm.Prices = map[abi.ChainEpoch]vm.Pricelist{
+		0: oldPrices[build.UpgradeHyggeHeight],
+	}
+	t.Cleanup(func() { vm.Prices = oldPrices })
+
+	client, _, ens := kit.EnsembleMinimal(t, kit.MockProofs())
+	ens.InterconnectAll().BeginMining(10 * time.Millisecond)
+
+	// First, try sending a message that should have no fees beyond the inclusion cost. I.e., it
+	// does absolutely nothing:
+	msg := &types.Message{
+		From:       client.DefaultKey.Address,
+		To:         client.DefaultKey.Address,
+		Value:      big.Zero(),
+		GasLimit:   0,
+		GasFeeCap:  abi.NewTokenAmount(10000),
+		GasPremium: big.Zero(),
+	}
+
+	burntBefore, err := client.WalletBalance(ctx, builtin.BurntFundsActorAddr)
+	require.NoError(t, err)
+	balanceBefore, err := client.WalletBalance(ctx, client.DefaultKey.Address)
+	require.NoError(t, err)
+
+	// Sign the message and compute the correct inclusion cost.
+	var smsg *types.SignedMessage
+	for i := 0; ; i++ {
+		var err error
+		smsg, err = client.WalletSignMessage(ctx, client.DefaultKey.Address, msg)
+		require.NoError(t, err)
+		estimatedGas := vm.PricelistByEpoch(math.MaxInt).OnChainMessage(smsg.ChainLength()).Total()
+		if estimatedGas == msg.GasLimit {
+			break
+		}
+		// Try 10 times to get the right gas value.
+		require.Less(t, i, 10, "unable to estimate gas: %s != %s", estimatedGas, msg.GasLimit)
+		msg.GasLimit = estimatedGas
```

Comment:
Do we loop because setting the gas limit makes the message longer, so we need to re-re-compute the limit to get correct one?

If yes, that probably deserves a code comment.

Reply:
Yes.